go/pkg/sysdb/coordinator/table_catalog.go

```diff
@@ -859,3 +942,149 @@ func (tc *Catalog) FlushCollectionCompaction(ctx context.Context, flushCollectio
+// 5. 		If version CAS fails - then fail the operation to the Compactor.
+// 6. 		If version file name CAS fails - read updated file and write a new version file to S3.
+func (tc *Catalog) FlushCollectionCompactionForVersionedCollection(ctx context.Context, flushCollectionCompaction *model.FlushCollectionCompaction) (*model.FlushCollectionInfo, error) {
+	// The result that is sent back to the Compactor.
+	flushCollectionInfo := &model.FlushCollectionInfo{
+		ID: flushCollectionCompaction.ID.String(),
+	}
+
+	// Do the operation in a loop until the CollectionEntry is updated,
+	// 		OR FAIL the operation if the version is stale
+	//      OR other DB error.
+	//
+	// In common case, the loop will run only once.
+	// The loop with run more than once only when GC competes to update the
+	// VersionFileName. More precisely, when GC updates the VersionFile in S3
+	// to mark certain versions and then tries to update the VersionFileName in
+	// the table at the same time.
+	for {
+		// Get the current version info and the version file from the table.
+		collectionEntry, err := tc.metaDomain.CollectionDb(ctx).GetCollectionEntry(types.FromUniqueID(flushCollectionCompaction.ID), nil)
+		if err != nil {
+			return nil, err
+		}
+		if collectionEntry == nil {
+			return nil, common.ErrCollectionNotFound
+		}
+		if collectionEntry.IsDeleted {
+			return nil, common.ErrCollectionSoftDeleted
+		}
+
+		versionAtCompactionStart := int64(flushCollectionCompaction.CurrentCollectionVersion)
+		existingVersion := int64(collectionEntry.Version)
+
+		// Do a check to see if the version is stale.
+		if existingVersion > versionAtCompactionStart {
+			// Compactor is trying to flush a version that is no longer valid, since
+			// a different compaction instance has already incremented the version.
+			log.Info("Compactor is trying to flush a stale version", zap.Int64("existing_version", existingVersion), zap.Int64("current_collection_version", versionAtCompactionStart))
+			return nil, common.ErrCollectionVersionStale
+		}
+
+		existingVersionFileName := collectionEntry.VersionFileName
+		// Read the VersionFile from S3MetaStore.
+		existingVersionFilePb, err := tc.s3Store.GetVersionFile(flushCollectionCompaction.TenantID, flushCollectionCompaction.ID.String(), existingVersion, existingVersionFileName)
+		if err != nil {
+			return nil, err
+		}
+
+		// The update function takes the content of the existing version file,
+		// and the set of segments that are part of the new version file.
+		// NEW VersionFile is created in S3 at this step.
+		newVersionFileName, err := tc.updateVersionFileInS3(ctx, existingVersionFilePb, flushCollectionCompaction)
+		if err != nil {
+			return nil, err
+		}
+
+		txErr := tc.txImpl.Transaction(ctx, func(txCtx context.Context) error {
+			// register files to Segment metadata
+			err = tc.metaDomain.SegmentDb(txCtx).RegisterFilePaths(flushCollectionCompaction.FlushSegmentCompactions)
```

Comment:
nit: Please add a comment that writing to the segments table always needs to be done before updating the collections table. As I understand with read committed isolation level, the following could happen if the order is reversed:

`
TX1 updates collections table
TX2 updates collections table
TX2 updates segments table
TX1 updates segments table
`
so we'd end up segments from TX2 and collections from TX1. Classic problem that Repeatable Read isolation level solves